Add `gitlab:` git source shorthand

jgarber623 opened the pull request:
First off, thank you for considering this Pull Request! I appreciate your time.

## What was the end-user or developer problem that led to this PR?

In #7210, I noted the existence of the `github:` and `bitbucket:` shorthands for installing gems from Git sources. I proposed adding two new providers, `gitlab:` and `codeberg:`, for the notable Git services [Gitlab](https://gitlab.com) and [Codeberg](https://codeberg.org). The proposal to add Gitlab was accepted.

The changes proposed in this PR would provide a small quality-of-life improvement for users with Ruby dependencies whose source repositories exist somewhere other than GitHub.

Resolves #7210.

## What is your fix for the problem, implemented in this PR?

This PR adds `gitlab:` shorthand with similar functionality to the existing `github:` shorthand:

```ruby
gem "gitlab-sdk", gitlab: "gitlab-org/analytics-section/product-analytics/gl-application-sdk-rb"
```

Merge Request URLs are also supported for each new shorthand:

```ruby
gem "gitlab-sdk", gitlab: "https://gitlab.com/gitlab-org/analytics-section/product-analytics/gl-application-sdk-rb/-/merge_requests/27"
```

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

deivid-rodriguez commented:
Thanks!

I think in the related issue there was pretty much consensus on adding Gitlab. I think codeberg was more unclear, though. In particular, thresholds proposed by @indirect at https://github.com/rubygems/rubygems/issues/7210#issuecomment-1839714042 would exclude it.

So we can proceed in 2 ways:
* Add Gitlab only for now.
* Wait a bit until we make a final decision.

Regarding the implementation, it's pretty confusing, but the tests you added would be for an implementation of the Bundler DSL that actually lives in RubyGems and it's not really used by Bundler at the moment. The proper place for these tests would be in `bundler/spec/bundler/dsl_spec.rb`.

indirect commented:
I think we should definitely add GitLab. I am less clear on whether we should add Codeberg (and Sourcehut, and Beanstalk, and CodeCommit, and Cloud Source Repositories, and Launchpad, and Keybase, and Planio, and Tuleap). My main hesitation adding smaller repository sources is churn, having to constantly do work to update our shortcuts if URLs change or services shut down or move.

If we see consistent user requests for other sources, I am fine with adding them. So far I think we have only seen that for GitLab.

deivid-rodriguez commented:
@jgarber623 Could you limit this PR to adding support for the `gitlab` shorthand?

jgarber623 commented:
Thanks for the feedback, y'all! I've reverted the commit adding the Codeberg shortcut. Let me know if you'd rather avoid the revert commit and I can force push. Not sure which of GitHub's many merge approaches you prefer.

@deivid-rodriguez My apologies for that first pass at the tests. I've added a bunch of `gitlab`-specific ones in 98905e6, basing them on the existing `github` specs (since the two shortcuts are mostly the same). Let me know if there's anything else worth testing.

Thanks again and appreciate your consideration!

deivid-rodriguez commented:
No problem at all! Not your fault, it's a bit confusing :sweat_smile: 

Yes, if you can squash everything into one commit, I do prefer that :+1:.

Thanks for the PR!

deivid-rodriguez reviewed (approved):
Thank you!

simi commented on bundler/lib/bundler/dsl.rb:
```diff
@@ -308,6 +309,20 @@ def add_git_sources
         "https://#{user_name}@bitbucket.org/#{user_name}/#{repo_name}.git"
       end
+      
+      git_source(:gitlab) do |repo_name|
```
Since "gitlab" is name of the project and it could be also self-hosted and this points to public instance of gitlab.com, should this be `gitlab_com`?

simi replied:
> Isn't that also true about GitHub enterprise?

GitHub EE is not so common per my experience, but gitlab is often self-hosted by organizations. I'm also not happy about github shortcut btw.

martinemde replied:
> I'm also not happy about github shortcut btw. 

I agree. We were moving in the way of declaring it within the Gemfile.

This PR is a small ask overall. We've now drawn a clear line and we're enforcing it. I think this will be the last addition for a while baring a major shift.

We should think about, and even purposely deprecate the others that are smaller. That will give us a pattern for deprecating these when they are too small. I'd accept one no deprecating bitbucket, provided it printed the exact code to add to your Gemfile to fix the deprecation.

martinemde replied:
Sorry, I misremembered the bitbucket traffic numbers. Scratch that. Anything smaller should be removed though.

simi replied:
> I think this will be the last addition for a while barring a major shift.

:shrug:

martinemde commented:
Waiting for tests to be fixed.

jgarber623 commented:
Formatting issues addressed. 👍🏻

Both `rspec ./spec/quality_spec.rb` and `rake rubocop` are reporting all green.

martinemde commented:
Thank @jgarber623 !
